lib/axon_onnx/deserialize.ex

```diff
@@ -658,6 +667,69 @@ defmodule AxonOnnx.Deserialize do
     {Map.put(axon, output_name, Axon.activation(inp, activation, name: output_name)), used_params}
   end
 
+  # Builds an Axon layer which returns a new layer with input values 
+  # concatenated on the given axis 
+  defp to_axon_concat(
+         %Node{attribute: attrs, input: inputs, output: [output_name]},
+         axon,
+         params,
+         used_params
+       ) when is_list(inputs) do
+    inputs = for inp <- inputs, do: input_or_param!(inp, params, axon, used_params)
+    %{ "axis" => axis } = options!(attrs)
+
+    {Map.put(axon, output_name, Axon.concatenate(inputs, [axis: axis, name: output_name])), used_params}
+  end
+
+  # Builds an Axon layer which returns a new layer upsampling the input
+  # layer. The scale activation layer must contain 1.0 as the first two values
+  # Each dimension value of the output layer is:
+  # output_dimension = floor(input_dimension * scale)
+  defp to_axon_upsample(
+         %Node{attribute: attrs, input: [inp, scale], output: [output_name]},
+         axon,
+         params,
+         used_params
+       ) do
+    inp = input_or_param!(inp, params, axon, used_params)
+    scale = input_or_param!(scale, params, axon, used_params)
+    updated_params = 
+      used_params
+      |> Map.put(output_name <> "_scale", scale)
+    unless [1.0, 1.0 | scale] = scale|>Nx.to_flat_list do
+      raise "Wrong scale format. First two values must be [1, 1]."
+    end
+    [_, _ | inp_output_shape] = Tuple.to_list(inp.output_shape)
```

Comment:
First, you can match directly on the input to extract shape with:

```elixir
%Axon{output_shape: shape} = inp = input_or_param!(...)
```

Then we can refactor:

```suggestion
    output_shape =
      case Nx.to_flat_list(scale) do
        [1.0, 1.0 | scale] ->
          output_shape =
            scale
            |> Enum.zip_with(shape, fn x, y -> floor(x * y) end)
            |> List.to_tuple()
   
        [s1, s2 | scale] ->
          raise ArgumentError, "unspported scale format, first two scale values must be 1, got #{s1} and #{s2}"
```

Reply:
Oh I see, thank you for this, it is far more elegant than mine :)